measure

```diff
@@ -87,6 +100,113 @@ insights_session = url_session(INSIGHTS_API_URL.format(account_id=NEWRELIC_ACCOU
 insights_session.headers.update({'X-Query-Key': NEWRELIC_INSIGHTS_QUERY_KEY})
 
 
+def aggregate_metric_values_to_scalar(
+    metric_name: str,
+    metrics: Dict[str, Any],
+    result: Dict[str, Any],
+    is_threshold: bool,
+) -> bool:
```

Comment:
(cosmetic)
wouldn't it be simpler to define this as (metric: Dict) -> (result:float, err:string)?

(rather than pass metrics, metric_name and a boolean that defines in an indirect way where to store the output in a 'result' dictionary)

Providing the name might have been useful, if it were used in preparing error messages, but it isn't (except for the check that name is present in metrics).

If this is for the purpose of 'constructing' error messages from bits and pieces, this is a practice that should be avoided (mind, I was guilty of it myself in the past on many occasions). Even if this stays as an English-only code, future changes can easily result in funny status messages. And if localization is ever needed, this will be rejected by translators as unworkable. It is best to use entire sentences and have replacement strings in them only for values that come from user data that must be quoted-as is, or are themselves entire sentences that can always fit into another sentence as a subordinate clause, such as "an error has occurred in computing FOO: {placehoder-for-error-sentence-here}".

BTW, same goes for using the python-expression-inside-a-text-string thingie (imported into directly from Perl hell, straight into Py3.7 and on). It is not something that you want people that don't read Python (such as proof-readers or translators) to see or mess with. IMO, "{{compute_this('mess')}" should not be used to format text intended for non-Pythonized humans.

Please consider changing the function signature and moving the "error context" formatting to the caller, such as "error in aggregating metric {stuff}': {error-text-from-aggregate-here}".

(like I said, not critical or urgent, functionality-wise this is fine)

Reply:
The result also [stores the relevant instance IDs](https://github.com/opsani/servo-newrelic/blob/48e0b9c49000134c7231952b1af667297a99acdc/measure#L140) that returned data for the query. This is printed to the driver's STDOUT in the full results payload which is functionally ignored by servo (no 'progress' key and will never be the last stdout line). However, if servo is set to verbose slo check results will show up in the servo container logs for debugging.

If this is overkill, I could remove the instance_ids storing from the result. Or I could have the inst IDs come back as a third field of the tuple (if tuples are larger than 2 I generally lean toward return by reference). LMK your thoughts.

WRT: message clause constructing, I hadn't considered your points and really appreciate the feedback. I have a cleanup branch where I nitpicked all the bad style I could find and the message construction is more in line with your suggestion but there are two lines that may still be too computation heavy. If time permits, could you give those lines a quick look and LMK if they could benefit from being organized differently?

- https://github.com/opsani/servo-newrelic/blob/dae35761ea0db8e4ff0fd3fb7a13d56251438e66/slo.py#L216
- https://github.com/opsani/servo-newrelic/blob/dae35761ea0db8e4ff0fd3fb7a13d56251438e66/slo.py#L182